R/class_jvm_stats.R

```diff
@@ -0,0 +1,307 @@
+        totMem      = "numeric",
+        freeMem     = "numeric",
+        allocMem    = "numeric",
+        memUnit     = "character"
+    ),
+    prototype = list(
+        nJars       = NA_integer_,
+        javaVersion = NA_character_,
+        phgVersion  = NA_character_,
+        maxMem      = NA_real_,
+        totMem      = NA_real_,
+        freeMem     = NA_real_,
+        allocMem    = NA_real_,
+        memUnit     = NA_character_
+    )
+)
+
+
+## ----
+#' @title Create an instance of the JvmStats class
+#'
+#' @description
+#' Collects JVM statistics such as the number of JAR files loaded, Java
+#' version, PHG version, and memory usage. Returns an instance of the
+#' \code{JvmStats} class populated with these statistics.
+#'
+#' @return An object of class \code{JvmStats}, containing the following
+#' fields:
+#' \itemize{
+#'   \item \code{nJars}: The number of JAR files in the JVM classpath.
+#'   \item \code{javaVersion}: The version of Java being used.
+#'   \item \code{phgVersion}: The version of PHG (Practical Haplotype
+#'   Graph) being used.
+#'   \item \code{maxMem}: The maximum memory (in GB) that the JVM can use.
+#'   \item \code{totMem}: The total memory (in GB) allocated to the JVM.
+#'   \item \code{freeMem}: The amount of free memory (in GB) in the JVM.
+#'   \item \code{allocMem}: The amount of allocated memory (in GB) in the
+#'   JVM.
+#' }
+#'
+#' @details
+#' This function requires that the JVM is initialized. If the JVM is not
+#' initialized, it will throw an error. Use the \code{initPhg()} function
+#' to initialize the JVM with the PHG library path before calling this
+#' function.
+#'
+#' The function retrieves the number of JAR files in the classpath, the
+#' version of Java and PHG being used, and memory statistics including
+#' maximum memory, total memory, free memory, and allocated memory (all in
+#' gigabytes).
+#'
+#' @examples
+#' \dontrun{
+#' # Ensure that the JVM is initialized with PHGv2 library path
+#' initPhg("/path/to/phgv2")
+#'
+#' # Collect JVM statistics
+#' stats <- jvmStats()
+#' show(stats)
+#' }
+#'
+#' @seealso \code{\link{JvmStats-class}}
+#'
+#' @export
+jvmStats <- function() {
+    if (!isJvmInitialized()) {
+        rlang::abort("JVM is not initialized - please run 'initPhg()' with PHGv2 library path")
+    }
+
+    # Get number of JARs in class path
+    classPath <- rJava::.jclassPath()
+    nJars <- length(classPath[grepl("\\.jar$", classPath)])
+
+    # Get Java version
+    javaVersion <- rJava::.jcall("java/lang/System", "S", "getProperty", "java.version")
+
+    # Attempt to get PHG version
+    genUtils <- rJava::J(PHG_JVM$GEN_UTILS)
+    if (!any(grepl("phgVersion()", names(genUtils)))) {
+        phgVersion <- paste0(cli::symbol$leq, " 2.4.1.155")
+    } else {
+        phgVersion <- genUtils$phgVersion()
+    }
+
+    # Get memory profile
+    runtime  <- rJava::.jcall("java/lang/Runtime", "Ljava/lang/Runtime;", "getRuntime")
+    gbConv   <- 1024 ^ 3
```

Comment:
Is there a reason to just not make `1073741824` the constant?

Reply:
Nope - just easier to remember $1024^{3}$ when I was coding this and forgot to update! Updated.